draft-ietf-quic-datagram.md

```diff
@@ -206,20 +206,32 @@ and can store the contents in memory.
 
 DATAGRAM frames MUST be protected with either 0-RTT or 1-RTT keys.
 
-Application protocols using datagrams are responsible for defining the
-semantics of the Datagram Data field, and how it is parsed. If the application
-protocol supports the coexistence of multiple entities using datagrams inside
-a single QUIC connection, it can use a mechanism to allow demultiplexing
-between them. For example, using datagrams with HTTP/3 involves prepending
-a flow identifier to all datagrams, see {{?I-D.schinazi-quic-h3-datagram}}.
-
 Note that while the max_datagram_frame_size transport parameter places a limit
 on the maximum size of DATAGRAM frames, that limit can be further reduced by
 the max_packet_size transport parameter, and by the Maximum Transmission Unit
 (MTU) of the path between endpoints. DATAGRAM frames cannot be fragmented,
 therefore application protocols need to handle cases where the maximum datagram
 size is limited by other factors.
 
+## Multiplexing Datagrams
+
+DATAGRAM frames belong to a QUIC connection as a whole, and are not strongly
+associated with any stream ID at the QUIC layer. However, it is expected
+that applications will want to differentiate between different DATAGRAM frames
+by using identifiers, such as for logical flows of datagrams or to distinguish
+between different kinds of datagrams.
+
+Identifiers used to multiplex different kinds of datagrams, or flows of
+datagrams, are the responsibility of the application protocol running over QUIC
+to define. The application defines the semantics of the Datagram Data field
+and how it is parsed.
+
+If the application needs to support the coexistence of multiple flows of
+datagrams, one recommended pattern is to use a variable-length integer at the
+beginning of the Datagram Data field. For example, using datagrams with
+HTTP/3 involves prepending such a flow identifier to all datagrams
+(see {{?I-D.schinazi-quic-h3-datagram}}).
```

Comment:
I assume H3 datagram spec also references back to this one? Does it make sense to have a circular reference? I know some in the IETF discussion wanted this text, but I'm personally fine with out this last paragraph.

Reply:
Agreed, though FWIW circular normative references are OK - they just require simultaneous publication.

Reply:
I'm not too bothered by a circular reference but I don't think it adds much. The wire format is simple. The main complexity of the H3 draft comes from how the flow IDs meld with HTTP semantics. So if you're going to reference it then be more specific about why, or just don't bother IMO.

And if you are going to reference it, let's use the MASQUE edited draft

Reply:
Given some of the churn in H3 DGRAM, how about we extract this third paragraph into another issue/PR that we can sit on while we land the other changes?

Reply:
@LPardue removed the specific H3 datagram reference; that OK?

Reply:
yep!